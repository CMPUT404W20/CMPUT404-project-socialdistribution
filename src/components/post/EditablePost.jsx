import React, { Component } from "react";
import "../../styles/post/EditablePost.scss";
import PropTypes from "prop-types";
import ImageOutlinedIcon from "@material-ui/icons/ImageOutlined";
import VisibilityRoundedIcon from "@material-ui/icons/VisibilityRounded";
import TextareaAutosize from "react-textarea-autosize";
import UploadImageModal from "./UploadImageModal";
import PostPreviewModal from "./PostPreviewModal";

class EditablePost extends Component {
  constructor(props) {
    super(props);
    this.state = {
      uploadModalVisibility: false,
      previewModalVisibility: false,
      originalPost: props.originalPost,
      postTitle: props.defaultPostTitle,
      postContent: props.defaultPostContent,
      postImage: props.defaultPostImage,
      postVisibility: "PUBLIC",
    };
  }

  handleTitleChange = (event) => {
    this.setState({ postTitle: event.target.value });
  };

  handleTitleKeyPress = (event) => {
    // disable the enter key so the user can't have multi-line comments
    // can still have long text that spans multiple lines though
    if (event.key === "Enter") {
      event.preventDefault();
    }
  }

  handleTextChange = (event) => {
    this.setState({ postContent: event.target.value });
  };

  handleImageUpload = (image) => {
    this.setState({ postImage: URL.createObjectURL(image) });
  }

  handleSubmit = (event) => {
    event.preventDefault();

    const {
      originalPost,
      postTitle,
      postContent,
      postImage,
      postVisibility,
    } = this.state;

    const { onSubmit } = this.props;

    originalPost.content = postContent;
    originalPost.imageSrc = postImage;
    originalPost.title = postTitle;
    originalPost.visibility = postVisibility;

    onSubmit(originalPost);
    this.setState({
      postTitle: "",
      postContent: "",
      postImage: "",
    });
    // eslint-disable-next-line no-alert
    // alert(postContent);
  };

  toggleUploadModalVisibility = () => {
    this.setState((prevState) => ({
      uploadModalVisibility: !prevState.uploadModalVisibility,
    }));
  }

  togglePreviewModalVisibility = () => {
    this.setState((prevState) => ({
      previewModalVisibility: !prevState.previewModalVisibility,
    }));
  }

  changePostVisibility = (event) => {
    this.setState({
      postVisibility: event.target.value,
    });
  }


  render() {
    const {
      uploadModalVisibility,
      previewModalVisibility,
      postTitle,
      postContent,
      postImage,
    } = this.state;

    const { editMode, onDiscard } = this.props;

    // Marcos, https://stackoverflow.com/questions/2476382/how-to-check-if-a-textarea-is-empty-in-javascript-or-jquery
    const titleLength = postTitle.replace(/^\s+|\s+$/g, "").length;
    const postLength = postContent.replace(/^\s+|\s+$/g, "").length;
    const validPost = titleLength > 0 && (postLength > 0 || postImage !== "");

    const componentTitle = editMode ? "EDIT POST" : "NEW POST";

    return (
      <div className="editable-post-wrapper">
        <div className="editable-post-content">
          <div className="editable-post-header">
            <b>{componentTitle}</b>
            <select className="privacy-select" defaultValue="PUBLIC" onChange={this.changePostVisibility}>
              <option value="PUBLIC">Anyone</option>
              <option value="PRIVATE">Specific author</option>
              <option value="FRIENDS">Friends</option>
              <option value="FOAF">Mutual friends</option>
              <option value="SERVERONLY">Server Only</option>
              <option value="UNLISTED">Private</option>
            </select>
          </div>
          <UploadImageModal
            show={uploadModalVisibility}
            onHide={this.toggleUploadModalVisibility}
            onUpload={this.handleImageUpload}
          />
          <PostPreviewModal
            show={previewModalVisibility}
            onHide={this.togglePreviewModalVisibility}
            postTitle={postTitle}
            postContent={postContent}
            imageObjectUrl={postImage}
          />
          <form className="editable-post-input-wrapper" action="submit">
            <TextareaAutosize
              placeholder="Title"
              className="title-text-area"
              onChange={this.handleTitleChange}
              onKeyPress={this.handleTitleKeyPress}
              value={postTitle}
            />
            <TextareaAutosize
              placeholder="What's on your mind?"
              className="post-text-area"
              onChange={this.handleTextChange}
              value={postContent}
            />
            {
              postImage ? (
                <img src={postImage} className="preview-image" alt="preview" />
              ) : null
            }
            <div className="editable-post-buttons-wrapper">
              {
                validPost ? (
                  <VisibilityRoundedIcon
                    className="icon"
                    onClick={this.togglePreviewModalVisibility}
                  />
                ) : null
              }

              <ImageOutlinedIcon
                className="icon"
                onClick={this.toggleUploadModalVisibility}
              />
              {
                editMode ? (
                  <button
                    type="submit"
                    className="discard-button"
                    onClick={onDiscard}
                  >
                    Discard
                  </button>
                ) : null
              }
              <button
                type="submit"
                className="post-button"
                onClick={this.handleSubmit}
                disabled={!validPost}
              >
                { editMode ? "Update" : "Post" }
              </button>
            </div>
          </form>
        </div>
      </div>
    );
  }
}

EditablePost.propTypes = {
  editMode: PropTypes.bool,
  originalPost: PropTypes.shape({
    id: PropTypes.string,
    title: PropTypes.string,
    authorId: PropTypes.string,
    username: PropTypes.string,
    published: PropTypes.string,
    imageSrc: PropTypes.string,
    content: PropTypes.string,
  }),
  defaultPostTitle: PropTypes.string,
  defaultPostContent: PropTypes.string,
  defaultPostImage: PropTypes.string,
  onSubmit: PropTypes.func.isRequired,
  onDiscard: PropTypes.func,
};

EditablePost.defaultProps = {
  editMode: false,
  originalPost: {},
  defaultPostTitle: "",
  defaultPostContent: "",
  defaultPostImage: "",
  onDiscard: () => {},
};

export default EditablePost;
