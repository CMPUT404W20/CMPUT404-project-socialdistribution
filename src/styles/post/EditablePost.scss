@import "../common.scss";

.editable-post-wrapper {
  background-color: white;
  box-sizing: border-box;
  box-shadow: $post-box-shadow;
  border-radius: $radius;
  width: 100%;
  display: flex;
  justify-content: center;

  .editable-post-content {
    display: flex;
    flex-direction: column;
    width: 100%;
    padding: 0 15pt;
  }

  .editable-post-header {
    display: flexbox;
    flex-direction: row;
    justify-content: center;
    border-bottom: 1px solid $border-color;
    height: fit-content;
    padding-top: 10pt;
    padding-bottom: 3pt;
  
    & > b {
      font-style: normal;
      font-weight: 500;
      font-size: 10pt;
      text-transform: uppercase;
      color: black;
    }
  
    .privacy-select {
      font-style: normal;
      font-size: 10pt;
      margin-left: 1.5rem;
    }  
  }

  .editable-post-input-wrapper {
    margin-top: 1.0rem;
    
    .title-text-area {
      width: 100%;
      min-height: 20px;
      font-size: 14pt;
      font-weight: 500;
      border: none;
      outline: 0;
      resize: none;
      font-style: italic;
      margin-bottom: -5pt;
    }

    .post-text-area {
      width: 100%;
      min-height: 80px;
      font-size: 12pt;
      border: none;
      outline: 0;
      resize: none;
    }

    .preview-image {
      width: 100%;
      margin-bottom: 20px;
    }
  }  

  .editable-post-buttons-wrapper {
    display: flex;
    float: right;
    padding-bottom: 15pt;
    
    .icon {
      @extend %icon;
      margin: 2px 0.2rem;
    }
  
    .post-button {
      @extend %button;
      margin-left: 0.8rem;
    }

    .discard-button {
      @extend %button;
      background-color: #fa93a7;
      &:not(:disabled) {
        &:hover, &:focus {
          outline: none;
          background-color: #e73a5a;
        }
      }
      margin-left: 0.8rem;
    }
  }
}
