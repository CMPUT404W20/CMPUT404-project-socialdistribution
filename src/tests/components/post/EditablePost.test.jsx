import React from "react";
import { shallow } from "enzyme";
import EditablePost from "../../../components/post/EditablePost";

describe("Post box Component", () => {
  it("should render correctly", () => {
    const component = shallow(
      <EditablePost
        onSubmit={() => {}}
      />,
    );
    expect(component).toMatchSnapshot();
  });
});
