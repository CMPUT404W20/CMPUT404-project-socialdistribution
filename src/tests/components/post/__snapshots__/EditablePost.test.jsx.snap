// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post box Component should render correctly 1`] = `
<div
  className="editable-post-wrapper"
>
  <div
    className="editable-post-content"
  >
    <div
      className="editable-post-header"
    >
      <b>
        NEW POST
      </b>
    </div>
    <UploadImageModal
      onHide={[Function]}
      onUpload={[Function]}
      show={false}
    />
    <PostPreviewModal
      imageObjectUrl=""
      onHide={[Function]}
      postContent=""
      postTitle=""
      show={false}
    />
    <PrivacySelectorModal
      onHide={[Function]}
      onSubmit={[Function]}
      selectedVisibileTo={Array []}
      selectedVisibility="PUBLIC"
      show={false}
    />
    <form
      action="submit"
      className="editable-post-input-wrapper"
    >
      <TextareaAutosize
        className="title-text-area"
        inputRef={[Function]}
        onChange={[Function]}
        onHeightChange={[Function]}
        onKeyPress={[Function]}
        placeholder="Title"
        useCacheForDOMMeasurements={false}
        value=""
      />
      <TextareaAutosize
        className="post-text-area"
        inputRef={[Function]}
        onChange={[Function]}
        onHeightChange={[Function]}
        placeholder="What's on your mind?"
        useCacheForDOMMeasurements={false}
        value=""
      />
      <div
        className="editable-post-buttons-wrapper"
      >
        <PublicIcon
          className="icon"
          onClick={[Function]}
        />
        <ImageOutlinedIcon
          className="icon"
          onClick={[Function]}
        />
        <button
          className="post-button"
          disabled={true}
          onClick={[Function]}
          type="submit"
        >
          Post
        </button>
      </div>
    </form>
  </div>
</div>
`;
